src/diagnostic/GraphProbe.cpp

```diff
@@ -0,0 +1,420 @@
+/** Copyright 2017-2021 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#include <tvm/diagnostic/GraphProbe.h>
+
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <string>
+
+namespace
+{
+using namespace tvm::diagnostic;
+using namespace tvm::graph::internal;
+
+std::string indent(int val) { return std::string(val, ' '); }
+
+// shared code for any and none (see below)
+template<bool b>
+bool check(const std::vector<GraphProbe::OutputVal> & outputVal, std::function<bool(const Eigen::MatrixXd &)> select)
+{
+  for(const auto & v : outputVal)
+  {
+    if(select(std::get<Eigen::MatrixXd>(v)))
+      return b;
+  }
+  return !b;
+}
+
+// any(outputVal, select) returns true if any value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto any = check<true>;
+// none(outputVal, select) returns true if no value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto none = check<false>;
```

Comment:
I think this is a little confusing, how about:

```cpp
template<auto Algorithm>
bool check(const std::vector<GraphProbe::OutputVal> & outputVal, const std::function<bool(const Eigen::MatrixXd &)> & select)
{
  return Algorithm(outputVal.begin(), outputVal.end(), [&select](const GraphProbe::OutputVal & v) { return select(std::get<Eigen::MatrixXd>(v)); });
}

using iterator_t = std::vector<GraphProbe::OutputVal>::const_iterator;
using predicate_t = const std::function<bool(const GraphProbe::OutputVal &)> &;
constexpr auto any = check<std::any_of<iterator_t, predicate_t>>;
constexpr auto none = check<std::none_of<iterator_t, predicate_t>>;

(or just writing `any` and `none` in terms of `std::any_of` and `std::none_of` without any template machinery involved)
```

Reply:
I adopted your proposed code.